include/eeprom/core/dev_data.hpp

```diff
@@ -0,0 +1,384 @@
+        if (tail_updated && config_updated) {
+            auto table_location = calculate_table_entry_start(key);
+            if (table_location > data_tail) {
+                LOG("Error, attemping to read uninitalized value");
+                return;
+            }
+            if (len > data.size()) {
+                LOG("ERROR, trying to write %d bytes from a %lu byte buffer",
+                    len, data.size());
+                len = data.size();
+            }
+            LOG("Writing %d bytes to data partition", len);
+            // copy the data to our internal buffer
+            std::copy_n(data.begin(), len, this->type_data.begin());
+            this->action_cmd_m =
+                table_entry_action{.key = key,
+                                   .offset = offset,
+                                   .len = len,
+                                   .action = TableAction::WRITE};
+            // call a read to the table entry so we know where
+            // to put the data
+            this->eeprom_client.send_eeprom_queue(message::ReadEepromMessage{
+                .memory_address = table_location,
+                .length = static_cast<types::data_length>(2 * conf.addr_bytes),
+                .callback = table_action_callback,
+                .callback_param = this});
+        }
+    }
+    template <std::size_t SIZE>
+    void write_data(uint16_t key, uint16_t len,
+                    std::array<uint8_t, SIZE>& data) {
+        write_data(key, len, 0, data);
+    }
+    template <std::size_t SIZE>
+    void write_data(uint16_t key, std::array<uint8_t, SIZE>& data) {
+        write_data(key, data.size(), 0, data);
+    }
+
+    void get_data(uint16_t key, uint16_t len, uint16_t offset) {
+        if (tail_updated && config_updated) {
+            auto table_location = calculate_table_entry_start(key);
+            if (table_location > data_tail) {
+                LOG("Error, attemping to read uninitalized value");
+                return;
+            }
+            action_cmd_m = table_entry_action{.key = key,
+                                              .offset = offset,
+                                              .len = len,
+                                              .action = TableAction::READ};
+
+            // call a read to the table entry so we know where
+            // to read the data
+            this->eeprom_client.send_eeprom_queue(message::ReadEepromMessage{
+                .memory_address = table_location,
+                .length = static_cast<types::data_length>(2 * conf.addr_bytes),
+                .callback = table_action_callback,
+                .callback_param = this});
+        }
+    }
+    void get_data(uint16_t key, uint16_t len) { get_data(key, len, 0); }
+    void get_data(uint16_t key) { get_data(key, 0, 0); }
+
+    template <std::size_t SIZE>
+    void create_data_part(uint16_t key, uint16_t len,
+                          std::array<uint8_t, SIZE>& data) {
```

Comment:
So the data being written to the eeprom will always need to be passed in as byte arrays since this is the way the underlying eeprom reads/writes data. Implementing clients will need to serialize/parse anything that they want to read/write.
This accessor aims to be as generic as possible. 

Also the data being passed in here is not of DataBufferType because the data being written can come from any type of byte data the client wants to send and it doesn't necessarily need to match the size of the DataBufferType used as the backing storage.